src/persist/src/indexed/compact.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Persistence compaction runtime.
+
+use std::sync::Arc;
+
+use differential_dataflow::lattice::Lattice;
+use differential_dataflow::trace::Description;
+use timely::progress::Antichain;
+use timely::PartialOrder;
+use tokio::runtime::Runtime;
+
+use crate::error::Error;
+use crate::future::Future;
+use crate::indexed::cache::BlobCache;
+use crate::indexed::encoding::{BlobTraceBatch, TraceBatchMeta};
+use crate::storage::Blob;
+
+/// A request to merge two trace batches and write the results to blob storage.
+pub struct CompactReq {
+    /// One of the batches to be merged.
+    pub b0: TraceBatchMeta,
+    /// One of the batches to be merged.
+    pub b1: TraceBatchMeta,
+    /// The since frontier to be used for the output batch. This must be at or
+    /// in advance of the since frontier for both of the input batch.
+    pub since: Antichain<u64>,
+    /// The blob key to write the resulting batch to.
+    pub output_key: String,
+}
+
+/// A successful compaction.
+pub struct CompactRes {
+    /// The original request.
+    pub req: CompactReq,
+    /// The compacted batch.
+    pub merged: TraceBatchMeta,
+}
+
+/// A runtime for asynchronous batch compaction.
+//
+// TODO: Add migrating records from unsealed to trace as well as deletion of
+// batches.
+pub struct Compacter<B: Blob> {
```

Comment:
I would just call this `Maintainer` because the other tasks in the todo all seem like background maintenance things to me

Reply:
Done in #8554